src/v/cluster/topic_table_partition_generator.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cluster/topic_table.h"
+
+namespace cluster {
+struct partition_replicas {
+    model::ntp partition;
+    std::vector<model::broker_shard> replicas;
+};
+
+class topic_table_partition_generator_exception : public std::runtime_error {
+public:
+    explicit topic_table_partition_generator_exception(const std::string& m)
+      : std::runtime_error(m) {}
+};
+
+/*
+ * This is a utility class that walks the topic table and returns batches
+ * of partitions and their replicas. For this operation to make sense,
+ * topic table stability is required. The generator will throw if the topic
+ * table has been updated between batches.
+ */
+class topic_table_partition_generator {
+public:
+    using generator_type_t = std::vector<partition_replicas>;
+
+    static constexpr size_t default_batch_size = 256;
+
+    explicit topic_table_partition_generator(
+      ss::sharded<topic_table>& topic_table,
+      size_t batch_size = default_batch_size)
+      : _topic_table(topic_table)
+      , _stable_revision_id(_topic_table.local().last_applied_revision())
+      , _batch_size(batch_size) {
+        if (_topic_table.local()._topics.empty()) {
+            _topic_iterator = _topic_table.local()._topics.end();
+            _exhausted = true;
+        } else {
+            _topic_iterator = _topic_table.local()._topics.begin();
+            _partition_iterator = current_assignment_set().begin();
+        }
+    }
+
+    ss::future<std::optional<generator_type_t>> next_batch() {
```

Comment:
nit: Could this be a function instead? It more models a function to me because after you call `next_batch()` one time `exhausted` is true, and the instance can never do any functional work again.

Maybe the class could be some sort of adapter that holds state that's passed to a function.

Reply:
Maybe I'm not reading your comment right, but you can call `next_batch` multiple times before `exhausted` becomes true. I think it needs to be a class because we have to keep state around between invocations of `next_batch`.